Add OpenStack Security Group module

emonty opened the pull request.

emonty committed:
Add OpenStack Security Group support

Two modules - one for security groups and one to manage rules in a
security group.

omgjlk commented on cloud/openstack/os_security_group.py:
```diff
@@ -0,0 +1,113 @@
+
+    try:
+        cloud = shade.openstack_cloud(**module.params)
+        nova_client = cloud.nova_client
```
Is nova client being used here to mask the difference between nova-network and neutron for the group creation?

omgjlk commented on cloud/openstack/os_security_group.py:
```diff
@@ -0,0 +1,113 @@
+        cloud = shade.openstack_cloud(**module.params)
+        nova_client = cloud.nova_client
+        changed = False
+        secgroup = cloud.get_security_group(module.params['name'])
```
Since security groups can have non-unique names, is there anything in shade that protects against finding multiple groups with the same name here?

omgjlk commented on cloud/openstack/os_security_group.py:
```diff
@@ -0,0 +1,113 @@
+        secgroup = cloud.get_security_group(module.params['name'])
+
+        if module.params['state'] == 'present':
+            secgroup = cloud.get_security_group(module.params['name'])
```
Why a second call to get the secgroup? This call just happened on line 82.

omgjlk commented on cloud/openstack/os_security_group.py:
```diff
@@ -0,0 +1,113 @@
+                                description=module.params['description'])
+                changed = True
+
+        if secgroup and secgroup.description != module.params['description']:
```
This functionality of being able to update the description should be demonstrated in the docstring, along with details that idempotence is on the name of group only.

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,156 @@
+options:
+   security_group:
+     description:
+        - Name of the security group
```
Can this be an ID as well, if we have the secgroup ID?

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,156 @@
+   remote_ip_prefix:
+      description:
+        - Source IP address(es) in CIDR notation (exclusive with remote_group)
+     required: false
```
If you don't supply a remote_ip_prefix, what is the default behavior?

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,156 @@
+        nova_client = cloud.nova_client
+        changed = False
+
+        secgroup = cloud.get_security_group(module.params['security_group'])
```
Same question as in the other module, what if there are multiple groups with the same name? Does shade bubble up an exception?

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,156 @@
+                                     rule=rule['id'])
+                changed = True
+
+        module.exit_json(changed=changed, result="success")
```
Might be friendly to return the security group rule ID with the return value too, for later attaching to an instance.

omgjlk replied:
Also shouldn't have result= here in the exit_json params.

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,156 @@
+                changed = True
+
+
+        if module.params['state'] == 'absent' and secgroup:
```
For some reason, Travis's python doesn't like this line.

omgjlk commented:
Completed review for this current offering. Some questions in the code, and then Travis can't compile the module even though the syntax looks right.

Commit:
Update for latest shade API

Shade version 0.7.0 introduces new API methods for creating,
deleting, and updating security groups. Let's use those and
clean up the module.

Commit:
Update the docstring for os_security_group

Indicate that idempotence is on security group names, and give
an example for updating a security group description.

Shrews commented:
@j2sol Just pushed a new version of os_security_group that addresses all of your comments. With the latest shade API that it now uses, an exception will be raised on duplicate security group names. And use of the nova client (or neutron client) is now hidden. Also updated the doctoring as you suggested.

omgjlk commented on cloud/openstack/os_security_group.py:
```diff
@@ -0,0 +1,145 @@
+        if state == 'present':
+            if not secgroup:
+                secgroup = cloud.create_security_group(name, description)
+                module.exit_json(changed=True, result='created',
```
I don't think it's proper to put a result= here. exit_json is always assumed a success, and the changed is used to indicate change. Adding a result here is confusing.

Shrews replied:
@j2sol I've removed all 'result=' values. These appear to be holdovers from the old module code and I agree that it's redundant/confusing.

omgjlk commented on cloud/openstack/os_security_group.py:
```diff
@@ -0,0 +1,145 @@
+                if _needs_update(module, secgroup):
+                    secgroup = cloud.update_security_group(
+                        secgroup['id'], description=description)
+                    module.exit_json(changed=True, result='updated',
```
Again shouldn't have a result=

omgjlk commented on cloud/openstack/os_security_group.py:
```diff
@@ -0,0 +1,145 @@
+                    module.exit_json(changed=True, result='updated',
+                                     id=secgroup['id'])
+                else:
+                    module.exit_json(changed=False, result='success')
```
success is assumed from exit_json vs fail_json.

omgjlk commented on cloud/openstack/os_security_group.py:
```diff
@@ -0,0 +1,145 @@
+
+        if state == 'absent':
+            if not secgroup:
+                module.exit_json(changed=False, result='success')
```
Same about result=

omgjlk commented:
Travis is still not happy with the syntax, and the group name vs group ID input for os_security_group_rule is not addressed. Other comments in-line.

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,156 @@
+                                     from_port=module.params['port_range_min'],
+                                     to_port=module.params['port_range_max'],
+                                     cidr=module.params['remote_ip'],
+                                     group_id=module.params['remote_group'],
```
This function call is missing a closing paren, which is causing the statement on line 141 to be invalid syntax.

Commit:
Remove 'result' value

This value is pretty much useless, and a holdover from the old
module code. Let's remove it.

Shrews commented:
@j2sol Agree with all your comments on the secgroup module, and the new commit has the fixes.

There is a fix in the PR pipeline for one of the @emonty branches for the rules module. The rules module is also due to be rewritten very soon anyway to rely more on shade and clean up the old module code.

Commit:
Remove os_security_group_rule module

The rules module will have it's own branch.

Shrews commented:
Removed os_security_group_rule module as it has it's own PR and branch.

omgjlk commented:
This now looks good to me! :+1:
